test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestClassLoader.java

```diff
@@ -2,38 +2,142 @@
         super(parent);
-        this.componentsProviderFile = Objects.requireNonNull(componentsProviderFile);
-        this.resourceReferenceProviderFile = resourceReferenceProviderFile;
+        this.name = name;
+        this.buildResult = buildResult;
+    }
+
+    @Override
+    public String getName() {
+        return "QuarkusComponentTestClassLoader: " + name;
+    }
+
+    public Map<String, Set<String>> getConfigMappings() {
+        return buildResult.configMappings();
+    }
+
+    public Map<String, String[]> getInterceptorMethods() {
+        return buildResult.interceptorMethods();
+    }
+
+    public Throwable getBuildFailure() {
+        return buildResult.failure();
+    }
+
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        synchronized (getClassLoadingLock(name)) {
+            Class<?> clazz = findLoadedClass(name);
+            if (clazz != null) {
+                return clazz;
+            }
+            byte[] bytecode = null;
+            if (buildResult.generatedClasses() != null) {
+                bytecode = buildResult.generatedClasses().get(name);
+                if (bytecode != null) {
+                    LOG.debugf("Use generated/transformed class for %s", name);
+                }
+            }
+            if (bytecode == null && !mustDelegateToParent(name)) {
+                bytecode = BYTECODE_CACHE.computeIfAbsent(name, this::loadBytecode).value();
+            }
+            if (bytecode != null) {
+                LOG.debugf("Define class %s", name);
+                clazz = defineClass(name, bytecode, 0, bytecode.length);
+                if (resolve) {
+                    resolveClass(clazz);
+                }
+                return clazz;
+            }
+            return super.loadClass(name, resolve);
+        }
+    }
+
+    private Bytecode loadBytecode(String name) {
+        byte[] bytecode = null;
+        String path = name.replace('.', '/') + ".class";
+        try (InputStream in = getParent().getResourceAsStream(path)) {
+            if (in != null) {
+                LOG.debugf("Loading class %s", name);
+                bytecode = in.readAllBytes();
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        return new Bytecode(bytecode);
+    }
+
+    private static boolean mustDelegateToParent(String name) {
+        return name.startsWith("java.")
+                || name.startsWith("jdk.")
+                || name.startsWith("javax.")
+                || name.startsWith("jakarta.")
+                || name.startsWith("sun.")
+                || name.startsWith("com.sun.")
+                || name.startsWith("org.w3c.")
+                || name.startsWith("org.xml.")
+                || name.startsWith("org.junit.")
+                || name.startsWith("org.mockito.")
+                || PARENT_CL_CLASSES.contains(name);
```

Comment:
Is there any way to get this code more common with the QuarkusClassLoader? I seem to recall doing a few [hacky overrides](https://github.com/quarkusio/quarkus/blob/f16dd61a3ab3bcc6c54ccc75ef74792f0b800cad/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java#L225) to what gets loaded parent-first over there to support dev services work. It might be (a) less code and (b) fewer confusing bugs if we had a common service.

Reply:
The list here looks quite close to what I had in #35473 and that wasn't a result of experimentation -- that was a list of packages provided by the JDK. There's a few packages added here (`jakarta.*`, `org.junit.*`, `org.mockito.*`) which make perfect sense, but still -- it's a list of packages we know are never provided by the application and make no sense to be loaded by the application class loader.

Reply:
> it's a list of packages we know are never provided by the application and make no sense to be loaded by the application class loader.

On the other hand, I observed weird errors while adding other packages which are also never provided by the app :shrug:.